### ISSUE TITLE
blitz/generator/generate Token Query Param in Cache

### ISSUE THREAD

**Issue by lukasNo1** [putyourlightson/craft-blitz#504]
Hello, i have a Problem with Blitz that sometimes occurs on the live environment. I wasn't able to reproduce it locally.

We use for example craft.app.request.absoluteUrl for the og:url meta tag. It adds a token to the url sometimes.

```
<meta property="og:url" content="https://www.website.com/site?token=60EGuPDAJwkS4XsXSi88RcnUfvYZjDXA">
```

If you would visit https://www.website.com/site the site has been fullpage cached with that token in it, regardless of the setting queryStringCaching => 1 which should "Cache URLs with query strings as unique pages"

I found this token in the "tokens" Table, it seems to be a blitz token.

token | route | usageLimit | usageCount | expiryDate | dateCreated | dateUpdated

60EGuPDAJwkS4XsXSi88RcnUfvYZjDXA | blitz/generator/generate | NULL | NULL | 2023-05-09 15:29:02 | 2023-05-08 15:29:02 | 2023-05-08 15:29:0

I do not want that token to be fullpage cached as this url will be invalid when the token expires.
Right now my quick fix is just to remove the "token" param from the craft.app.request.absoluteUrl in twig.

Can you make sense of this / figure out what the problem could be?

Blitz Config:
```
      'cachingEnabled'          => true,
      'clearCacheAutomatically' => true,
      'warmCacheAutomatically'  => false,
      'queryStringCaching'      => 1,
      'cacheDuration'           => 0,
      'includedUriPatterns'     => [
          [
              'siteId'     => '',
              'uriPattern' => '.*',
          ],
      ],
      'excludedUriPatterns'     => [],
```
Craft CMS 4.4.5
Blitz | 4.4.4

**Comment by bencroker** [comment]
The token parameter you’re seeing is what Blitz uses to generate the cache. 

The use of `craft.app.request.absoluteUrl` should be avoided as it can introduce cache poisoning, as you are experiencing above. Replacing it with something like `{{ url(craft.app.request.path) }}` should have the desired effect without the risk of including the token parameter.

I’ll also look into whether it might make sense for Blitz to ensure that token parameters do not exist in URLs when deciding whether or not to cache pages.

**Comment by lukasNo1** [comment]
Thank you for your response.

A check for the token before caching would be nice, because we have some use cases where we want to have the query string in cache, just not the token.

I can work around that for now and just remove the token param with a custom filter.

**Comment by bencroker** [comment]
The only problem with not caching pages that contain links with token parameters is that it is possible that a large portion of the site will not be cached via the cache generation job. I’ll have to investigate this further, but it sounds like you’ve found a workaround in the meantime.

**Comment by uandco** [comment]
I'm getting weird token issues too: I'm noticing a lot of 500s with this token parameter for URLs that don't exist.

For example `https://domain/some-invalid-url?token=7I3A459KL5jdiW3e60oK-UpVuTl56hwB` returns a 500:

```
2023-07-13 04:19:08 [web.ERROR] [TypeError] TypeError: putyourlightson\blitz\services\GenerateCacheService::save(): Argument #1 ($content) must be of type string, null given, called in /app/vendor/putyourlightson/craft-blitz/src/services/CacheRequestService.php on line 413 and defined in /app/vendor/putyourlightson/craft-blitz/src/services/GenerateCacheService.php:418
Stack trace:
#0 /app/vendor/putyourlightson/craft-blitz/src/services/CacheRequestService.php(413): putyourlightson\blitz\services\GenerateCacheService->save(NULL, Object(putyourlightson\blitz\models\SiteUriModel))
#1 /app/vendor/putyourlightson/craft-blitz/src/Blitz.php(307): putyourlightson\blitz\services\CacheRequestService->saveAndPrepareResponse(Object(craft\web\Response), Object(putyourlightson\blitz\models\SiteUriModel))
#2 [internal function]: putyourlightson\blitz\Blitz->putyourlightson\blitz\{closure}(Object(yii\base\Event))
#3 /app/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Object(Closure), Object(yii\base\Event))
#4 /app/vendor/yiisoft/yii2/base/Component.php(642): yii\base\Event::trigger('yii\\web\\Respons...', 'afterPrepare', Object(yii\base\Event))
#5 /app/vendor/yiisoft/yii2/web/Response.php(340): yii\base\Component->trigger('afterPrepare')
#6 /app/vendor/yiisoft/yii2/base/Application.php(390): yii\web\Response->send()
#7 /app/web/index.php(26): yii\base\Application->run()
#8 {main} {"memory":3821296,"exception":"[object] (TypeError(code: 0): putyourlightson\\blitz\\services\\GenerateCacheService::save(): Argument #1 ($content) must be of type string, null given, called in /app/vendor/putyourlightson/craft-blitz/src/services/CacheRequestService.php on line 413 at /app/vendor/putyourlightson/craft-blitz/src/services/GenerateCacheService.php:418)"} 
```

```
MySQL [uu-saslab-master_gtyq8]> select * from tokens where token = "7I3A459KL5jdiW3e60oK-UpVuTl56hwB";
+-----+----------------------------------+--------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
| id  | token                            | route                    | usageLimit | usageCount | expiryDate          | dateCreated         | dateUpdated         | uid                                  |
+-----+----------------------------------+--------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
| 574 | 7I3A459KL5jdiW3e60oK-UpVuTl56hwB | blitz/generator/generate |       NULL |       NULL | 2023-07-14 02:17:20 | 2023-07-13 02:17:20 | 2023-07-13 02:17:20 | 5fb3dd0f-5475-49f5-87d2-75bd92d0abe0 |
+-----+----------------------------------+--------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
```

I don't exactly understand how it ends up querying the page with the token in the URL, nor why it gets a 500.

Also "Flush cache" from the control panel doesn't clear this token, which is a bit puzzling.

I'm using Blitz 4.4.6 with Craft 4.4.15.

I'm also using Retour 4.1.12, not sure if it has any impact. The URL in question was not part of any static redirect configured in Retour and should have just delivered a 404.

Once I deleted the token from the database manually, it displayed the 404 page.

If I try to access any page with one of the tokens with the `blitz/generator/generate` route, I get the same 500:

```
select * from tokens;
+-----+----------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
| id  | token                            | route                                                                                                                                     | usageLimit | usageCount | expiryDate          | dateCreated         | dateUpdated         | uid                                  |
+-----+----------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
| 569 | tXg2bFrlEgw011XGONdYGtIK6RnhdKbw | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-13 05:37:11 | 2023-07-12 05:37:11 | 2023-07-12 05:37:11 | fcf688bf-261c-43b9-b3b8-00ae30149d6d |
| 570 | RMC2gfpG2ITp0DZPZE6X_GgrAmPbu0qy | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-13 05:46:51 | 2023-07-12 05:46:51 | 2023-07-12 05:46:51 | bb645664-2f86-4b5d-be0b-22faca1e8574 |
| 571 | UcsfR4nVER6b1loREonKYYdYOlT09iGu | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-14 01:16:45 | 2023-07-13 01:16:45 | 2023-07-13 01:16:45 | fcdebd63-ca66-43fb-9b5b-c0ce3b429169 |
| 572 | 3cjGDd4VZO8utifg29_M3Btl4QQnlK5Z | ["preview\/preview",{"elementType":"craft\\elements\\Entry","canonicalId":162,"siteId":1,"draftId":null,"revisionId":null,"userId":6540}] |       NULL |       NULL | 2023-07-14 02:10:54 | 2023-07-13 02:10:54 | 2023-07-13 02:10:54 | 9f01b55a-949f-460f-9e1b-e13a9cad00b0 |
| 573 | 0-LWub8-pXA7HjmbvS-xFptpja7T9TxA | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-14 02:12:58 | 2023-07-13 02:12:58 | 2023-07-13 02:12:58 | 7c89f6df-a1b1-4113-8fb4-2728771a18e9 |
| 575 | RI493U_Q97vvJPGkD-vmmNZ6toX41Kzw | ["preview\/preview",{"elementType":"craft\\elements\\Entry","canonicalId":600,"siteId":1,"draftId":null,"revisionId":null,"userId":6540}] |       NULL |       NULL | 2023-07-14 02:17:48 | 2023-07-13 02:17:48 | 2023-07-13 02:17:48 | fa1b97a6-a6ae-4c64-acd7-eb80b2cfbdd9 |
| 576 | 9K7eOb7CSvkUvkGY3KsxEJGPnYegrxhg | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-14 02:19:04 | 2023-07-13 02:19:04 | 2023-07-13 02:19:04 | ddd9bca8-8fdc-4c4d-8888-c5a16e885fb3 |
| 577 | N8Vl2SEGXXxJ6xlUkcPlsnLx8W3krPEo | blitz/generator/generate                                                                                                                  |       NULL |       NULL | 2023-07-14 02:59:36 | 2023-07-13 02:59:36 | 2023-07-13 02:59:36 | 8412ffad-3ffa-4b13-a73b-169b07544247 |
+-----+----------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+------------+------------+---------------------+---------------------+---------------------+--------------------------------------+
8 rows in set (0.003 sec)
```
All the 500s come from the server itself with local IPs and `amphp/http-client @ v4.x`

I'm also noticing 500s on URLs like `/shortcuts/contact-us`. `shortcuts` is a section that would point to 404s too as it doesn't have a dedicated template.

So, I reckon the automatic cache generation fails when trying to generate cache entries for resources which URL would result in a 404.

**Comment by bencroker** [comment]
This has already been addressed for the upcoming version 4.5.0, but since we’ve had a few delays on that release, I’ve backported it to the `develop` branch in https://github.com/putyourlightson/craft-blitz/commit/a672f7f0c137ac1dbcf011aca6840a5914b4fa80. 

You can test it by running `composer require putyourlightson/craft-blitz:dev-develop as 4.4.7`.
